models/sklearn/classification/classifierGRID.py

```diff
@@ -0,0 +1,153 @@
+                                )
+
+            self.logger.debug("Shape of X before train: {}".format(features_prepared.shape))
+            self.logger.info("Fitting the data to the model..")
+            gsvc.fit(features_prepared, self.y)
+
+            # print(gsvc.cv_results_["params"])
+            self.logger.info("Results from each best preprocess training:")
+            self.logger.info("a) Best score: {}".format(gsvc.best_score_))
+            self.logger.info("b) Best estimator: {}".format(gsvc.best_estimator_))
+            self.logger.info("c) Best parameters: {}".format(gsvc.best_params_))
+            self.logger.info("Counted evaluations in this GridSearch process: {}".format(len(gsvc.cv_results_["params"])))
+
+            # save best results for each train process
+            exports_dir = "{}_{}".format(self.config.get("exports_directory"), self.class_name)
+            results_path = FindCreateDirectory(self.exports_path,
+                                               os.path.join(exports_dir, "results")).inspect_directory()
+            results_best_dict_name = "result_{}_{}_best_{}.json"\
+                .format(self.class_name, tr_process["preprocess"], gsvc.best_score_)
+
+            results_dict = dict()
+            results_dict["score"] = gsvc.best_score_
+            results_dict["params"] = gsvc.best_params_
+            results_dict["n_fold"] = tr_process['n_fold']
+            results_dict["preprocessing"] = tr_process["preprocess"]
+            with open(os.path.join(results_path, results_best_dict_name), 'w') as grid_best_json:
+                json.dump(results_dict, grid_best_json, indent=4)
+
+            # export parameters that the
+            results_params_dict_name = "result_{}_{}_params_{}.json"\
+                .format(self.class_name, tr_process["preprocess"], gsvc.best_score_)
+            with open(os.path.join(results_path, results_params_dict_name), 'w') as grid_params_json:
+                json.dump(gsvc.cv_results_["params"], grid_params_json, indent=0)
+
+            models_path = FindCreateDirectory(self.exports_path,
+                                              os.path.join(exports_dir, "models")).inspect_directory()
+            best_process_model_path = os.path.join(models_path, "model_grid_{}.pkl".format(tr_process["preprocess"]))
+            joblib.dump(gsvc.best_estimator_, best_process_model_path)
+            self.logger.info("Grid Best model for the {} process saved.".format(tr_process["preprocess"]))
+
+            # return a list that includes the best models exported from each processing
+            self.best_models_list.append(results_dict)
+
+            print(colored("Next train process..", "yellow"))
+            process_counter += 1
+            print()
+            print()
+        print(colored("Finishing training processes..", "blue"))
+        print()
+
+    def export_best_classifier(self):
+        # gather best scores from the exported grid clf models
```

Comment:
in gaia we output a results file that gives the accuracy and parameters of each item in the grid search. Do we do something similar here?

Reply:
Here, the outcome of this function is the best model from all the best models that are exported in each preprocessing step (basic, normalize, etc.). Gaia has a similar output too. Furthermore, the best model from each preprocessing step is also extracted, as well as all the params combinations that are executed for that step.
* best model params and score: https://github.com/tzamalisp/acousticbrainz-server/blob/sklearn_integration/models/sklearn/classification/classifier_grid.py#L107-L114
* All the grid params combinations executed for each step: https://github.com/tzamalisp/acousticbrainz-server/blob/sklearn_integration/models/sklearn/classification/classifier_grid.py#L116-L120